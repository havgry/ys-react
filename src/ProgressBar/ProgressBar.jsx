import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import '../base.css'
import '@youseedk/dna/css/elements/ys-progress.css'

const ProgressBar = (props) => {
  const {
    value,
    valueMin,
    valueMax,
    variant,
    className,
  } = props

  const wrapperProps = {
    className: classNames(
      'ys-progress',
      variant && `ys-progress--${variant}`,
      className,
    ),
  }

  const progress = Math.round((value - valueMin) / (valueMax - valueMin) * 100)
  const progressPercentage = `${progress}%`
  const label = `${progressPercentage} gennemf√∏rt`

  return (
    <div
      role="progressbar"
      aria-valuenow={value}
      aria-valuemin={valueMin}
      aria-valuemax={valueMax}
      aria-valuetext={label}
      {...wrapperProps}
    >
      <div
        className="ys-progress__bar"
        aria-hidden="true"
      >
        <div
          className="ys-progress__progress"
          style={{ width: progressPercentage }}
        >
          <span className="ys-u-visuallyhidden">
            {label}
          </span>
        </div>
      </div>
    </div>
  )
}

ProgressBar.propTypes = {
  value: PropTypes.number.isRequired,
  valueMin: PropTypes.number.isRequired,
  valueMax: PropTypes.number.isRequired,
  variant: PropTypes.oneOf(['light']),
  className: PropTypes.string,
}

ProgressBar.defaultProps = {
  variant: null,
  className: '',
}

export default ProgressBar
