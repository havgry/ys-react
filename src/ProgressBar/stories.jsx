import React from 'react'
import { storiesOf } from '@storybook/react'

import { text, select, number } from '@storybook/addon-knobs'

import ProgressBar from './ProgressBar'

const variantOptions = {
  default: null,
  light: 'light',
}

storiesOf('ProgressBar', module)
  .addParameters({ jest: ['ProgressBar'] })
  .add('Default', () => (
    <ProgressBar
      value={number('value', 40, {}, 'Content')}
      valueMin={number('valueMin', 0, {}, 'Content')}
      valueMax={number('valueMax', 100, {}, 'Content')}
      variant={select('variant', variantOptions, variantOptions.default, 'Visuals')}
      className={text('className', '', 'Misc')}
    />
  ))
