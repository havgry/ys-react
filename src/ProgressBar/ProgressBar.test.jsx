import React from 'react'
import { shallow, mount as mountComponent } from 'enzyme'

import ProgressBar from './ProgressBar'

const customClassName = 'badger'
const getProgressBar = props => (
  <ProgressBar
    value={50}
    valueMin={0}
    valueMax={100}
    {...props}
  />
)

const shallowRender = props => shallow(getProgressBar(props))
const mount = props => mountComponent(getProgressBar(props))

describe('<ProgressBar>', () => {
  describe('Props', () => {
    it('Throws propType warning if no `value` prop is passed', () => {
      const component = () => shallowRender({
        value: null,
      })
      expect(component).toThrow()
    })

    it('Throws propType warning if no `valueMin` prop is passed', () => {
      const component = () => shallowRender({
        valueMin: null,
      })
      expect(component).toThrow()
    })

    it('Throws propType warning if no `valueMax` prop is passed', () => {
      const component = () => shallowRender({
        valueMax: null,
      })
      expect(component).toThrow()
    })

    it('Throws propType warning if `value` props aren\'t numbers', () => {
      let component = () => shallowRender({
        value: '10',
      })
      expect(component).toThrow()

      component = () => shallowRender({
        valueMin: '0',
      })
      expect(component).toThrow()

      component = () => shallowRender({
        valueMax: '100',
      })
      expect(component).toThrow()
    })

    it('Throws propType warning if unsupported `variant` prop is passed', () => {
      const component = () => shallowRender({
        variant: 'dark',
      })

      expect(component).toThrow()
    })

    it('Has correct defaultProps', () => {
      const component = mount()

      const {
        variant,
        className,
      } = component.props()

      expect(variant).toBe(null)
      expect(className).toBe('')
    })
  })

  describe('Rendering', () => {
    it('Correctly calculates progress', () => {
      const component = shallowRender()

      expect(component.find('.ys-progress__progress').prop('style')).toEqual({
        width: '50%',
      })
    })

    it('Correctly calculates progress with rounding', () => {
      const component = shallowRender({
        value: 1337,
        valueMin: 42,
        valueMax: 9000,
      })

      expect(component.find('.ys-progress__progress').prop('style')).toEqual({
        width: '14%',
      })
    })
  })

  describe('Snapshots', () => {
    it('With custom `className`', () => {
      const component = shallowRender({
        className: customClassName,
      })

      expect(component).toMatchSnapshot()
    })

    it('Without custom `className`', () => {
      const component = shallowRender()

      expect(component).toMatchSnapshot()
    })

    it('`light` variant', () => {
      const component = shallowRender({
        variant: 'light',
      })

      expect(component).toMatchSnapshot()
    })
  })
})
