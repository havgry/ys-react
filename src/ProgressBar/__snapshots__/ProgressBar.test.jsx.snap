// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar> Snapshots \`light\` variant 1`] = `
<div
  aria-valuemax={100}
  aria-valuemin={0}
  aria-valuenow={50}
  aria-valuetext="50% gennemført"
  className="ys-progress ys-progress--light"
  role="progressbar"
>
  <div
    aria-hidden="true"
    className="ys-progress__bar"
  >
    <div
      className="ys-progress__progress"
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <span
        className="ys-u-visuallyhidden"
      >
        50% gennemført
      </span>
    </div>
  </div>
</div>
`;

exports[`<ProgressBar> Snapshots With custom \`className\` 1`] = `
<div
  aria-valuemax={100}
  aria-valuemin={0}
  aria-valuenow={50}
  aria-valuetext="50% gennemført"
  className="ys-progress badger"
  role="progressbar"
>
  <div
    aria-hidden="true"
    className="ys-progress__bar"
  >
    <div
      className="ys-progress__progress"
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <span
        className="ys-u-visuallyhidden"
      >
        50% gennemført
      </span>
    </div>
  </div>
</div>
`;

exports[`<ProgressBar> Snapshots Without custom \`className\` 1`] = `
<div
  aria-valuemax={100}
  aria-valuemin={0}
  aria-valuenow={50}
  aria-valuetext="50% gennemført"
  className="ys-progress"
  role="progressbar"
>
  <div
    aria-hidden="true"
    className="ys-progress__bar"
  >
    <div
      className="ys-progress__progress"
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <span
        className="ys-u-visuallyhidden"
      >
        50% gennemført
      </span>
    </div>
  </div>
</div>
`;
